daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -33,81 +34,82 @@ import org.apache.daffodil.util.Maybe
 import org.apache.daffodil.util.Maybe._
 import org.apache.daffodil.util.MaybeULong
 import org.apache.daffodil.util.Misc
+import org.apache.daffodil.util.Logging
 import java.io.File
 
 /**
  * This simple extension just gives us a public method for access to the underlying byte array.
  * That way we don't have to make a copy just to access the bytes.
  */
-private[io] class ByteArrayOrFileOutputStreamWithGetBuf(
-  chunkSizeInBytes: Long = 4096,
-  maxBufferSizeInBytes: Long = 2000 * (1 << 20))
-  extends java.io.ByteArrayOutputStream {
-  private var _byteArrayOS = this
-  private var _fileOS: java.io.FileOutputStream = null
-
-  var isFile: Boolean = false
+private[io] class ByteArrayOutputStreamWithGetBuf() extends java.io.ByteArrayOutputStream {
+  def getCount = count
+  def getBuf = buf
+}
+
+private[io] class ByteArrayOrFileOutputStream(
+  maxBufferSizeInBytes: Long,
+  tempFilePath: Path,
+  useFileOnly: Boolean)
+  extends java.io.OutputStream
+  with Logging {
+  val filePath = {
+    if (tempFilePath.toString == "")
+      Paths.get(System.getProperty("java.io.tmpdir"))
+    else
+      tempFilePath
+  }
+
+  var stream: java.io.OutputStream = {
+    if (useFileOnly)
+      new java.io.FileOutputStream(File.createTempFile("daffodil-", ".tmp", filePath.toFile))
+    else
+      new ByteArrayOutputStreamWithGetBuf()
+  }
+
+  var isFile: Boolean = useFileOnly
   var tempFile: File = null
-  var bytesWritten = count
-
-  def switchOS() = {
-    Assert.invariant(_fileOS == null)
-    tempFile = File.createTempFile("daffodil-", ".tmp")
-    _fileOS = new java.io.FileOutputStream(tempFile)
-    this.flush
-    this.writeTo(_fileOS)
-    bytesWritten = count
-    this.reset
-    isFile = true
+  var nBytes: Long = 0
+
+  @inline
+  private def writeBytes(lengthInBytes: Long): Unit = {
+    if (!isFile && (nBytes + lengthInBytes > maxBufferSizeInBytes)) {
+      log(LogLevel.Info, "Switching to file based output stream. If this is performance critical, you may want to consider re-organizing your schema to avoid this if possible.")
+      tempFile = File.createTempFile("daffodil-", ".tmp", filePath.toFile)
+      val newStream: java.io.OutputStream = new java.io.FileOutputStream(tempFile)
+      stream.flush
+      stream.asInstanceOf[ByteArrayOutputStreamWithGetBuf].writeTo(newStream)
+      stream = newStream
+      isFile = true
+    }
+    nBytes += lengthInBytes
   }
 
   override def write(b: Array[Byte]) = {
-    if ((_fileOS == null) && ((count + b.length) > maxBufferSizeInBytes))
-      switchOS()
-
-    if (_fileOS != null) {
-      _fileOS.write(b)
-      bytesWritten += b.length
-    } else
-      super.write(b)
+    writeBytes(b.length)
```

Comment:
writeBytes feels like an odd name, considering it doesn't always write bytes. Only writes bytes from the buffer to the file if switching. Not sure of a different name though.